onnxruntime/core/session/inference_session.cc

```diff
@@ -1826,6 +1827,45 @@ Status InferenceSession::PartialRun(onnxruntime::RunOptions& run_options,
 }
 #endif
 
+namespace {
+// Concurrent runs counting and thread-pool spin control
+struct ThreadPoolSpinningSwitch {
+  using PS = onnxruntime::concurrency::ThreadPool::ParallelSection;
+  concurrency::ThreadPool* intra_tp_{nullptr};
+  concurrency::ThreadPool* inter_tp_{nullptr};
+  std::atomic<int>& concurrent_num_runs_;
+  // Use this to jump start threads and amortize the costs
+  // of initialization between the kernels
+  // note this prevents using explicit PS in the nodes
+  // or an additional PS for inter op thread-pool
+  std::optional<PS> ps_intra_;
+  // __Ctor ref-counting only
+  explicit ThreadPoolSpinningSwitch(std::atomic<int>& ref) : concurrent_num_runs_(ref) {
+    concurrent_num_runs_.fetch_add(1, std::memory_order_relaxed);
+  }
+  // __Ctor Refcounting and spinning control
+  ThreadPoolSpinningSwitch(concurrency::ThreadPool* intra_tp,
+                           concurrency::ThreadPool* inter_tp,
+                           std::atomic<int>& ref) noexcept
+      : intra_tp_(intra_tp), inter_tp_(inter_tp), concurrent_num_runs_(ref) {
+    if (concurrent_num_runs_.fetch_add(1, std::memory_order_relaxed) == 0) {
+      if (intra_tp_) intra_tp_->EnableSpinning();
+      if (inter_tp_) inter_tp_->EnableSpinning();
+    }
```

Comment:
Do we need an additional level of counting here to handle global thread pools, or is that handled above?  i.e.

- InferenceSession counts the number of concurrent run requests, and calls DisableSpinning once it is idle
- A global thread pool shared by multiple inference sessions needs to add its own counting and only actually disable spinning once _all_ of the inference sessions using it have gone idle?

Reply:
This change re-uses the same counter, but in an exception safe manner.
Comment about global thread pool is noted. I am not going to touch that one.